File: app/javascript/channels/chat_room_channel.js
import consumer from "./consumer"

const chatRoomChannel = consumer.subscriptions.create("ChatRoomChannel", {
  connected() {
    console.log("Connected to the chat room!");
    $("#messages").scrollTop($("#messages")[0].scrollHeight);
    $("#modal").css('display', 'flex');
  },

  disconnected() {

  },

  received(data) {
    if (data.message) {
      let current_name = sessionStorage.getItem('chat_room_name')
      let msg_class = data.sent_by === current_name ? "sent" : "received"
      $('#messages').append(`<p class='${msg_class}'> <em>`+data.sent_by+'</em> <br>' + data.message + '</p>')
      $("#messages").scrollTop($("#messages")[0].scrollHeight);
    } else if(data.chat_room_name) {
      let name = data.chat_room_name;
      let announcement_type = data.type == 'join' ? 'joined' : 'left';
      $('#messages').append(`<p class="announce"><em>${name}</em> ${announcement_type} the room</p>`)
      $("#messages").scrollTop($("#messages")[0].scrollHeight);
    }
  },

  speak(message) {
    let name = sessionStorage.getItem('chat_room_name')
    this.perform('speak', { message, name })
  },

  announce(content) {
    this.perform('announce', { name: content.name, type: content.type })
  }
});

export default chatRoomChannel;

File: app/channels/chat_room_channel.rb
class ChatRoomChannel < ApplicationCable::Channel
  def subscribed
    stream_from "chat_room_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def speak(data)
    data_type= 'message'
    @message_log = MessageLog.new(username: data['name'],body: data['message'], jenis: data_type)
    if @message_log.save
      ActionCable.server.broadcast "chat_room_channel", message: data["message"], sent_by: data["name"]
    end
  end

  def announce(data)
    data_type= 'announce'
    @message_log = MessageLog.new(username: data['name'],body: data['type'], jenis: data_type)
    if @message_log.save
      ActionCable.server.broadcast "chat_room_channel", chat_room_name: data["name"], type: data["type"]
    end
  end
end